File: android/libapp/src/main/java/com/topdon/ble/GenericRequest.java
package com.topdon.ble;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.topdon.ble.callback.RequestCallback;

import java.util.Queue;
import java.util.UUID;

class GenericRequest implements Request, Comparable<GenericRequest> {
    Device device;
    private final String tag;
    RequestType type;
    UUID service;
    UUID characteristic;
    UUID descriptor;
    Object value;
    int priority;
    RequestCallback callback;
    WriteOptions writeOptions;
    byte[] descriptorTemp;

    Queue<byte[]> remainQueue;
    byte[] sendingBytes;

    GenericRequest(RequestBuilder builder) {
        tag = builder.tag;
        type = builder.type;
        service = builder.service;
        characteristic = builder.characteristic;
        descriptor = builder.descriptor;
        priority = builder.priority;
        value = builder.value;
        callback = builder.callback;
        writeOptions = builder.writeOptions;
    }

    @Override
    public int compareTo(GenericRequest other) {
        return Integer.compare(other.priority, priority);
    }

    @NonNull
    public Device getDevice() {
        return device;
    }

    @NonNull
    public RequestType getType() {
        return type;
    }

    @Nullable
    public String getTag() {
        return tag;
    }

    @Nullable
    public UUID getService() {
        return service;
    }

    @Nullable
    public UUID getCharacteristic() {
        return characteristic;
    }

    @Nullable
    public UUID getDescriptor() {
        return descriptor;
    }

    @Override
    public void execute(Connection connection) {
        if (connection != null) {
            connection.execute(this);
        }
    }
}

File: android/libapp/src/main/java/com/topdon/ble/GenericRequest.kt
package com.topdon.ble

import androidx.annotation.NonNull
import androidx.annotation.Nullable
import com.topdon.ble.callback.RequestCallback
import java.util.*

internal class GenericRequest(builder: RequestBuilder) : Request, Comparable<GenericRequest> {
    var device: Device? = null
    private val tag: String? = builder.tag
    val type: RequestType = builder.type
    val service: UUID? = builder.service
    val characteristic: UUID? = builder.characteristic
    val descriptor: UUID? = builder.descriptor
    val value: Any? = builder.value
    val priority: Int = builder.priority
    val callback: RequestCallback? = builder.callback
    val writeOptions: WriteOptions? = builder.writeOptions
    var descriptorTemp: ByteArray? = null
    
    var remainQueue: Queue<ByteArray>? = null
    var sendingBytes: ByteArray? = null
    
    override fun compareTo(other: GenericRequest): Int {
        return other.priority.compareTo(priority)
    }
    
    @NonNull
    override fun getDevice(): Device {
        return device!!
    }
    
    @NonNull
    override fun getType(): RequestType {
        return type
    }
    
    @Nullable
    override fun getTag(): String? {
        return tag
    }
    
    @Nullable
    override fun getService(): UUID? {
        return service
    }
    
    @Nullable
    override fun getCharacteristic(): UUID? {
        return characteristic
    }
    
    @Nullable
    override fun getDescriptor(): UUID? {
        return descriptor
    }
    
    override fun execute(connection: Connection?) {
        connection?.execute(this)
    }
}